### PR TITLE
Qt6 py312

### PR DESCRIPTION
**Pull request by looooo**
Checklist
* [ ] Used a [personal fork of the feedstock to propose changes](https://conda-forge.org/docs/maintainer/updating_pkgs.html#forking-and-pull-requests)
* [ ] Bumped the build number (if the version is unchanged)
* [ ] Reset the build number to `0` (if the version changed)
* [ ] [Re-rendered]( https://conda-forge.org/docs/maintainer/updating_pkgs.html#rerendering-feedstocks ) with the latest `conda-smithy` (Use the phrase <code>@<space/>conda-forge-admin, please rerender</code> in a comment in this PR for automated rerendering)
* [ ] Ensured the license file is being packaged.

### CODE REVIEW

File: recipe/conda_build_config.yaml
Comment (looooo) [inline recipe/conda_build_config.yaml:1]: @conda-forge/core any ideas how to make the MACOSX-specification optional for py312/qt6?

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:1]: I don't quite understand. Can you rephrase this?

**Comment by hmaarrfk** [comment]
@conda-forge-admin please rerender

File: recipe/conda_build_config.yaml
Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:1]: if you mean

```
    - qt6-main  # [py >= 312]
```

can be adapted to
```
    - qt6-main  # [py >= 312 and not osx]
```

should work

Comment (looooo) [reply recipe/conda_build_config.yaml:1]: Sry, no I meant these lines which I have updated to work with qt6:
```
MACOSX_DEPLOYMENT_TARGET:
  - '10.15'
MACOSX_SDK_VERSION:
  - '11.0'
  ```
Not sure if these newer versions introduce issues for older macosx versions.

Comment (hmaarrfk) [reply recipe/conda_build_config.yaml:1]: see https://conda-forge.org/docs/maintainer/knowledge_base.html#requiring-newer-macos-sdks

I believe I added the line

**Comment by looooo** [comment]
Thanks @hmaarrfk. Before I can merge this I wait for another update of coin because the previous patch release added some abi-incompatibilities for windows https://github.com/coin3d/coin/issues/513
Also there is one failing build here: https://github.com/coin3d/pivy/pull/108#issuecomment-1826898594

**Comment by adrianinsaval** [comment]
anything holding this back?

**Comment by looooo** [comment]
@adrianinsaval not sure how to deal with this alpha release.

**Comment by looooo** [comment]
@adrianinsaval I created a new branch for the 0.6.9.a0 version. To install a build one can use this command:
```
conda install pivy -c conda-forge/label/pivy_rc
```

**Comment by oursland** [comment]
Any progress on this?  I have been toying with a `conda` built version of FreeCAD, but it is incomplete due to `pivy` not being available for Qt6 / Python 3.12.

**Comment by oursland** [comment]
I have added a draft PR for FreeCAD to add a Python 3.12 and Qt6 configuration here: https://github.com/FreeCAD/FreeCAD/pull/14218

It currently builds, although some dependencies have been installed via `apt` (`libegl1`, `libgl1`, `libglx0`, `libopengl0`), and passes all C++ tests and most Python tests, with exception to those that depend upon `pivy`.

@looooo @adrianinsaval

File: recipe/conda_build_config.yaml [hunk 1]
```diff
@@ -1,2 +1,6 @@
+MACOSX_SDK_VERSION:
+  - '11.0'
 vc:
-  - 14
\ No newline at end of file
+  - 14
+qt6_main:
+  - '6.6'
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml]: I think you'll need qt 6.7; pyside6 hasn't been building for 6.6 since over half a year, which is where the conflicts are coming from AFAICT
```suggestion
qt6_main:
  - '6.7'
```

File: recipe/conda_build_config.yaml [hunk 2]
```diff
@@ -1,2 +1,6 @@
+MACOSX_SDK_VERSION:
+  - '11.0'
 vc:
-  - 14
\ No newline at end of file
+  - 14
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml]: By the way, the `vc: 14` can be dropped (and please add a trailing newline as is standard for git-based repos)

File: recipe/conda_build_config.yaml [hunk 3]
```diff
@@ -1,6 +1,5 @@
 MACOSX_SDK_VERSION:
   - '11.0'
-vc:
-  - 14
 qt6_main:
-  - '6.6'
\ No newline at end of file
+  - '6.7'
+
```
Comment (h-vetinari) [inline recipe/conda_build_config.yaml]: Final line needs to be empty. ;-)

**Comment by looooo** [comment]
@conda-forge/core 
I tried clang for windows but running into these issues:
```
C:\Windows\system32\cmd.exe /C "cd . && D:\bld\pivy_1727261720886\_build_env\Library\bin\clang++.exe -nostartfiles -nostdlib -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fstack-protector-strong -O2 -ffunction-sections -pipe -D_CRT_SECURE_NO_WARNINGS -nostdlib -fms-runtime-lib=dll -fuse-ld=lld -fno-aligned-allocation  -O0 -D_DEBUG -D_DLL -D_MT -Xclang --dependent-lib=msvcrtd -g -Xclang -gcodeview -nostdlib -Wl,-defaultlib:D:/bld/pivy_1727261720886/_build_env/lib/clang/18/lib/windows/clang_rt.builtins-x86_64.lib   -Xlinker /subsystem:console  -fuse-ld=lld-link CMakeFiles/cmTC_f7d2f.dir/testCXXCompiler.cxx.obj -o cmTC_f7d2f.exe -Xlinker /MANIFEST:EMBED -Xlinker /implib:cmTC_f7d2f.lib -Xlinker /pdb:cmTC_f7d2f.pdb -Xlinker /version:0.0   -lkernel32 -luser32 -lgdi32 -lwinspool -lshell32 -lole32 -loleaut32 -luuid -lcomdlg32 -ladvapi32 -loldnames  && cd ."
    lld-link: error: could not open 'oldnames.lib': no such file or directory
    lld-link: error: could not open 'msvcrt.lib': no such file or directory
```
Is there a known fix for this?

**Comment by h-vetinari** [comment]
I've [asked](https://github.com/conda-forge/freecad-feedstock/pull/122#issuecomment-2257113852) you to please avoid pinning cf/core about the clang compilers, especially on non-standard platforms.

These libraries should be in the image; potentially something changed there that we need to update. Please open an issue on https://github.com/conda-forge/clang-win-activation-feedstock/.

**Comment by looooo** [comment]
> I've [asked](https://github.com/conda-forge/freecad-feedstock/pull/122#issuecomment-2257113852) you to please avoid pinning cf/core about the clang compilers, especially on non-standard platforms.

Sorry, I will try to remember this.

**Comment by looooo** [comment]
> These libraries should be in the image; potentially something changed there that we need to update. Please open an issue on https://github.com/conda-forge/clang-win-activation-feedstock/.

https://github.com/conda-forge/clang-win-activation-feedstock/issues/45

**Comment by looooo** [comment]
linux_64: `find_package(Python REQUIRED COMPONENTS Interpreter Development)` is failing for python3.12

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,39 +1,48 @@
...
     - swig
   host:
     - python
     - coin3d
-    - soqt
-    - qt-main
+    - soqt      # [py < 312]
+    - qt-main   # [py < 312]
+    - soqt6     # [py >= 312]
+    - qt6-main  # [py >= 312]
```
Comment (h-vetinari) [inline recipe/meta.yaml]: Same question as before: is it really the intention to switch to qt6 only for py>=312? This is going to lead to weird situations in other feedstocks and the solver.

Is there a technical reason why python 3.9-3.11 cannot similarly switch?

File: recipe/yum_requirements.txt [hunk 5]
```diff
@@ -1,4 +1,5 @@
 libXt-devel
 libXmu-devel
 libXi-devel
-mesa-libGLU-devel
\ No newline at end of file
+mesa-libGLU-devel
+mesa-libGL
```
Comment (h-vetinari) [inline recipe/yum_requirements.txt]: It would be good to get rid of (all) this in favour of `libglvnd(-devel)` resp the various xorg packages.

**Review by h-vetinari — COMMENTED** [review]
I see this is pointing to a new upstream version now, which is great. As such, I took a quick look.

File: recipe/meta.yaml (hunk 4 above)
Comment (looooo) [reply recipe/meta.yaml]: python 3.12 is failing with qt5 as pyside2 is not available for python3.12. But you are right, let's build the full build-matrix (qt5/qt6 x python3.9-python3.12)

File: recipe/yum_requirements.txt (hunk 5 above)
Comment (looooo) [reply recipe/yum_requirements.txt]: I think we should start with this in the coin3d-feedstock:
https://github.com/conda-forge/coin3d-feedstock

**Comment by h-vetinari** [comment]
Windows builds are passing now; the only failure is on aarch with CMake not finding numpy; you might find some help/inspiration in https://github.com/conda-forge/conda-forge.github.io/issues/2260 and references therein.

**Comment by looooo** [comment]
Is pyside2 for python3.12 now available. I expected these builds to fail.

packages are skipped, but ci-files are created with conda-smithy rerender.

**Comment by looooo** [comment]
> Windows builds are passing now; the only failure is on aarch with CMake not finding numpy; you might find some help/inspiration in [conda-forge/conda-forge.github.io#2260](https://github.com/conda-forge/conda-forge.github.io/issues/2260) and references therein.

not sure I understand correctly, but numpy is not a dependency of pivy.

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by h-vetinari** [comment]
Don't manually remove CI files. They don't really hurt as long as the skips are done correctly, and the risk of breaking something by manual intervention is not worth it

If you want to ensure these jobs do not get created by conda-smithy, you need to ensure that _all_ outputs (including the recipe-global `build:` section) have the corresponding skip present

**Comment by hmaarrfk** [comment]
ps you should rerendering when changing the yum_requirements.txt file

**Comment by looooo** [comment]
@conda-forge-admin, please rerender

**Comment by looooo** [comment]
Why x86_64 dependencies (yum) are pulled for cross compiled linuxaarch64 ?

**Comment by hmaarrfk** [comment]
> Why x86_64 dependencies (yum) are pulled for cross compiled linuxaarch64 ?

Because you are cross compiling.

```
build_platform:
  linux_aarch64: linux_64
```

**Comment by h-vetinari** [comment]
Nice, well done. Next stop freecad. ;)
